test/framework/finalizers_helpers.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterctlcluster "sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// finalizerAssertion contains a list of expected Finalizers corresponding to a resource Kind.
+var finalizerAssertion = map[string][]string{
+	"Cluster":             {clusterv1.ClusterFinalizer},
+	"Machine":             {clusterv1.MachineFinalizer},
+	"MachineSet":          {clusterv1.MachineSetTopologyFinalizer},
+	"MachineDeployment":   {clusterv1.MachineDeploymentTopologyFinalizer},
+	"ClusterResourceSet":  {addonsv1.ClusterResourceSetFinalizer},
+	"DockerMachine":       {infrav1.MachineFinalizer},
+	"DockerCluster":       {infrav1.ClusterFinalizer},
+	"KubeadmControlPlane": {controlplanev1.KubeadmControlPlaneFinalizer},
+}
+
+// ValidateFinalizersResilience checks that expected finalizers are in place, deletes them, and verifies that expected finalizers are properly added again.
+func ValidateFinalizersResilience(ctx context.Context, proxy ClusterProxy, namespace, clusterName string) {
+	clusterKey := client.ObjectKey{Namespace: namespace, Name: clusterName}
+
+	// Collect all objects where finalizers were initially set
+	objectsWithFinalizers := getObjectsWithFinalizers(ctx, proxy, namespace)
+
+	// Removes all the Finalizers.
+	// We are testing the worst-case scenario, i.e. all finalizers are deleted.
+	// The reconciliation loop should be able to recover from this, by adding the required Finalizers back.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, true)
```

Comment:
Q: Is it necessary to pause the Cluster?

Reply:
I'm asking because I think if we don't pause the cluster, removing the finalizers will trigger a reconcile which then will add the finalizers again. Which means we don't need the force reconciles

Reply:
Not necessary AFAIK, but it will make sure all reconciliations are paused while we remove all the finalizers.

Reply:
Yeah - this was inherited from the ownerReference tests where it's not strictly necessary either - but it may be more needed there as some reconcilers could return errors if the ownerRefs are in an inconsistent state.

I'm fine to either remove it or keep it as is. Keeping the pause / unpause makes it more of an atomic change which I can see the benefit in for trying to debug this.

Reply:
Sounds fine to me. Let's please document this in godoc here.